src/Microsoft.OData.Core/ODataBatchReader.cs

```diff
@@ -108,12 +95,54 @@ public ODataBatchReaderState State
                 return this.batchReaderState;
             }
 
-            private set
+            protected set
             {
                 this.batchReaderState = value;
             }
         }
 
+        /// <summary>
+        /// >The batch-specific URL converter that stores the content IDs found in a changeset and supports resolving cross-referencing URLs.
+        /// </summary>
+        internal ODataBatchPayloadUriConverter PayloadUriConverter
+        {
+            get { return this.payloadUriConverter; }
+        }
+
+        /// <summary>
+        /// Reader's input context.
+        /// </summary>
+        protected ODataInputContext InputContext
+        {
+            get { return this.inputContext; }
+        }
+
+        /// <summary>
+        /// Previously cache contentId that should be applied to the next message read.
+        /// </summary>
+        protected string ContentIdToAddOnNextRead
+        {
+            get { return this.contentIdToAddOnNextRead; }
+            set { this.contentIdToAddOnNextRead = value; }
+        }
+
+        /// <summary>
+        /// The reader's Operation state
+        /// </summary>
+        protected OperationState ReaderOperationState

```

Comment:
>ReaderOperationState [](start = 33, length = 20)

After moving ReadImplementation to the base class, this can (and should) be made private.  It is currently called only from ReadImplementation and from CreateOperation[Request|Response]Implementation; you can instead move the call to the CreateOperation[Request|Response]Message methods in the base class.

Reply:
Taking CreateOperationRequestMessageImplementation on  on 6.x branch PR as example, JsonLight implementation ([https://github.com/biaol-odata/odata.net/blob/maintenance-6.x_JsonFormatBatch-A/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchReader.cs#L108](https://github.com/biaol-odata/odata.net/blob/maintenance-6.x_JsonFormatBatch-A/src/Microsoft.OData.Core/JsonLight/ODataJsonLightBatchReader.cs#L108)) is completely different from the MultipartMixed implemenation ([https://github.com/biaol-odata/odata.net/blob/maintenance-6.x_JsonFormatBatch-A/src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs#L121](https://github.com/biaol-odata/odata.net/blob/maintenance-6.x_JsonFormatBatch-A/src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchReader.cs#L121)): JsonLight format has artifacts including atomicGroup, dependsOn with different processing logic. I am not seeing feasible common parts for these two implementations to achieve an abstract layer.

---
In reply to: [140929947](https://github.com/OData/odata.net/pull/960#discussion_r140929947) [](ancestors = 140929947)

Reply:
If we are able to do a clean refactoring, this should be able to be made private.

---
In reply to: [142581820](https://github.com/OData/odata.net/pull/960#discussion_r142581820) [](ancestors = 142581820,140929947)